include/aws/s3/private/s3_client_impl.h

```diff
@@ -76,9 +76,18 @@ struct aws_s3_endpoint {
     /* Callback for when this endpoint completely shuts down. */
     aws_s3_endpoint_shutdown_fn *shutdown_callback;
 
-    /* True, if the endpoint is created by client. So, it need to be thread safe to manage the refcount via
-     * `aws_s3_client_endpoint_release` */
-    bool handled_by_client;
+    /* For S3 client to handle the hashtable of endpoints. */
+    struct {
+        /* Point to the client. */
+        struct aws_s3_client *client;
+        /* If set, the release of endpoint will be scheduled to run within the task to prevent the data race with client
+         */
+        struct aws_task *task;
```

Comment:
why create a new task just for this? why not re-use the client's existing `process_work_task`?

we could keep an aws_array_list of `struct aws_s3_endpoint*` to be released